File: leetcode/Python/two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        us, dq, node = set(), collections.deque(), None
        dq.append(root)
        while dq:
            node = dq.popleft()
            if (k-node.val) in us:
                return True
            us.add(node.val)
            if node.left:
                dq.append(node.left)
            if node.right:
                dq.append(node.right)
        return False

File: leetcode/C++/two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        unordered_set<int> us;
        TreeNode* node=nullptr;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()){
            node=q.front(); q.pop();
            if(us.find(k-node->val)!=us.end()) return true;
            else us.insert(node->val);
            if(node->left) q.push(node->left);
            if(node->right) q.push(node->right);
        }
        return false;
    }
};